examples/core/Graphormer/train.py

```diff
@@ -0,0 +1,482 @@
+import argparse
+import random
+
+import torch as th
+import torch.nn as nn
+import torch.nn.functional as F
+from accelerate import Accelerator
+from dgl import shortest_dist
+from dgl.data import download
+from dgl.dataloading import GraphDataLoader
+from dgl.nn import DegreeEncoder, GraphormerLayer, PathEncoder, SpatialEncoder
+from ogb.graphproppred import DglGraphPropPredDataset, Evaluator
+from torch.nn.utils.rnn import pad_sequence
+from transformers.optimization import (
+    AdamW,
+    get_polynomial_decay_schedule_with_warmup,
+)
+
+
+class Graphormer(nn.Module):
+    def __init__(
+        self,
+        num_classes=1,
+        edge_dim=3,
+        num_atoms=4608,
+        max_degree=512,
+        num_spatial=511,
+        multi_hop_max_dist=5,
+        num_encoder_layers: int = 12,
+        embedding_dim: int = 768,
+        ffn_embedding_dim: int = 768,
+        num_attention_heads: int = 32,
+        dropout: float = 0.1,
+        pre_layernorm: bool = True,
+        activation_fn=th.nn.GELU(),
+    ):
+
+        super().__init__()
+        self.dropout = nn.Dropout(p=dropout)
+        self.embedding_dim = embedding_dim
+        self.num_heads = num_attention_heads
+
+        # 1 for graph token
+        self.atom_encoder = nn.Embedding(
+            num_atoms + 1, embedding_dim, padding_idx=0
+        )
+        self.graph_token = nn.Embedding(1, embedding_dim)
+
+        # degree encoder
+        self.degree_encoder = DegreeEncoder(
+            max_degree=max_degree, embedding_dim=embedding_dim
+        )
+
+        # path encoder
+        self.path_encoder = PathEncoder(
+            max_len=multi_hop_max_dist,
+            feat_dim=edge_dim,
+            num_heads=num_attention_heads,
+        )
+
+        # spatial encoder
+        self.spatial_encoder = SpatialEncoder(
+            max_dist=num_spatial, num_heads=num_attention_heads
+        )
+        self.graph_token_virtual_distance = nn.Embedding(1, num_attention_heads)
+
+        self.emb_layer_norm = nn.LayerNorm(self.embedding_dim)
+
+        self.layers = nn.ModuleList([])
+        self.layers.extend(
+            [
+                GraphormerLayer(
+                    feat_size=self.embedding_dim,
+                    hidden_size=ffn_embedding_dim,
+                    num_heads=num_attention_heads,
+                    dropout=dropout,
+                    activation=activation_fn,
+                    norm_first=pre_layernorm,
+                )
+                for _ in range(num_encoder_layers)
+            ]
+        )
+
+        # map graph_rep to num_classes
+        self.lm_head_transform_weight = nn.Linear(
+            self.embedding_dim, self.embedding_dim
+        )
+        self.layer_norm = nn.LayerNorm(self.embedding_dim)
+        self.activation_fn = activation_fn
+        self.embed_out = nn.Linear(self.embedding_dim, num_classes, bias=False)
+        self.lm_output_learned_bias = nn.Parameter(th.zeros(num_classes))
```

Comment:
Can we use a linear layer with "bias=True" instead?

Reply:
Since Graphormer resets the output parameters with uniform weight and **zero** bias when fine-tuning, the bias will be reset in uniform way if we just use a linear layer with "bias=True".

Reply:
Got it, thanks.